File: objc/Project3/Practice1/Views/SourceViewController.m
//
//  SourceViewController.m
//  Practice1
//
//  Created by Jinwoo Kim on 11/18/20.
//

#import "SourceViewController.h"
#import "DetailViewController.h"

@interface SourceViewController ()
@property (strong) IBOutlet NSTableView *tableView;
@property NSMutableArray<NSString *> *pictures;
@end

@implementation SourceViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do view setup here.
    NSFileManager *fm = [NSFileManager defaultManager];
    NSString *path = NSBundle.mainBundle.resourcePath;
    NSArray<NSString *> *items = [fm contentsOfDirectoryAtPath:path error:nil];
    self.pictures = [@[] mutableCopy];
    
    for (NSString *item in items) {
        @autoreleasepool {
            if ([item hasPrefix:@"nssl"]) [self.pictures addObject:item];
        }
    }
}

- (NSInteger)numberOfRowsInTableView:(NSTableView *)tableView {
    return [self.pictures count];
}

- (NSView *)tableView:(NSTableView *)tableView viewForTableColumn:(NSTableColumn *)tableColumn row:(NSInteger)row {
    NSTableCellView *vw = [tableView makeViewWithIdentifier:tableColumn.identifier owner:self];
    vw.textField.stringValue = self.pictures[row];
    return vw;
}

- (void)tableViewSelectionDidChange:(NSNotification *)notification {
    if (self.tableView.selectedRow == -1) return; // -1이면 아무것도 선택 안 됐다는 뜻
    NSSplitViewController *splitVC = self.parentViewController;
    DetailViewController *detail = splitVC.childViewControllers[1];
    [detail imageSelected:self.pictures[self.tableView.selectedRow]];
}

@end

File: swift/Project3/Project1/Views/SourceViewController.swift
//
//  SourceViewController.swift
//  Project1
//
//  Created by Jinwoo Kim on 11/18/20.
//

import Cocoa

class SourceViewController: NSViewController, NSTableViewDataSource, NSTableViewDelegate {
    @IBOutlet var tableView: NSTableView!
    var pictures = [String]()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do view setup here.
        let fm = FileManager.default
        let path = Bundle.main.resourcePath!
        let items = try! fm.contentsOfDirectory(atPath: path)
        
        for item in items {
            if item.hasPrefix("nssl") {
                pictures.append(item)
            }
        }
    }
    
    func numberOfRows(in tableView: NSTableView) -> Int {
        return pictures.count
    }
    
    func tableView(_ tableView: NSTableView,
                    viewFor tableColumn: NSTableColumn?,
                    row: Int) -> NSView? {
        guard let vw = tableView.makeView(withIdentifier: tableColumn!.identifier, owner: self) as? NSTableCellView else { return nil }
        vw.textField?.stringValue = pictures[row]

        return vw
    }
    
    func tableViewSelectionDidChange(_ notification: Notification) {
        guard tableView.selectedRow != -1 else { return } // -1이면 아무것도 선택 안 됐다는 뜻
        guard let splitVC = parent as? NSSplitViewController else { return }
        
        if let detail = splitVC.children[1] as? DetailViewController {
            detail.imageSelected(name: pictures[tableView.selectedRow])
        }
    }
}
